src/redisearch_rs/trie_rs/src/ffi.rs

```diff
@@ -181,8 +186,37 @@ unsafe extern "C" fn TrieMap_Add(
         debug_assert!(!str.is_null(), "str cannot be NULL if len > 0");
     }
 
-    let _unused = (value, cb);
-    todo!()
+    // SAFETY: The safety requirements of this function
+    // require the caller to ensure that the pointer `t` is
+    // a valid TrieMap obtained from `NewTrieMap` and cannot be NULL.
+    // If that invariant is upheld, then the following line is sound.
+    let TrieMap(trie) = unsafe { &mut *t };
+
+    let key = if len > 0 {
+        // SAFETY: The safety requirements of this function
+        // require the caller to ensure that the pointer `str` is
+        // a valid pointer to a C string, with a length of `len` bytes.
+        // If that invariant is upheld, then the following line is sound.
+        unsafe { slice::from_raw_parts(str, len as usize) }
+    } else {
+        &[]
+    };
+
+    let value = match cb {
+        Some(cb) => {
+            if let Some(old) = trie.find(key) {
+                // SAFETY: The following line is sound if the cb implementation
+                // does not introduce undefined behavior.
+                unsafe { cb(*old, value) }
```

Comment:
Can `cb` free `old`? It's not clear to me reading the function documentation.

Reply:
It is. Clarified in 30793852dbaa5ec7c6de3ddffe3eb6e30d9fa62b.